### PR TITLE
Add spec for RETURNDATASIZE opcode

### PR DESCRIPTION
**Pull request by ashWhiteHat**
# Abstract
I added RETURNDATASIZE opcode and modified missing link.

I would appreciate it if you could confirm.
Thank you.

### CODE REVIEW

File: tests/evm/test_returndatasize.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    bytecode = Bytecode().returndatasize()
+    bytecode_hash = RLC(bytecode.hash(), randomness)
+
+    tables = Tables(
+        block_table=set(),
+        tx_table=set(),
+        bytecode_table=set(bytecode.table_assignments(randomness)),
+        rw_table=set(
+            RWDictionary(9)
+            .call_context_read(1, CallContextFieldTag.CallDataLength, returndatasize)
```
Comment (xiaodino) [inline tests/evm/test_returndatasize.py:35]: `CallContextFieldTag.CallDataLength` should be `CallContextFieldTag.LastCalleeReturnDataLength`, right?

Comment (CPerezz) [reply tests/evm/test_returndatasize.py:35]: Agree!

File: specs/opcode/3dRETURNDATASIZE.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+   - `u64` is on top of stack
+
+## Exceptions
+
+1. stack overflow: stack is full, stack pointer = 0
+2. out of gas: remaining gas is not enough
+
+## Code
+
+Please refer to `src/zkevm_specs/evm/execution/returndatasize.py`.
```
Comment (CPerezz) [inline specs/opcode/3dRETURNDATASIZE.md:34]:
```suggestion
Please refer to [returndatasize.py](src/zkevm_specs/evm/execution/returndatasize.py).
```

File: specs/opcode/3dRETURNDATASIZE.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+### EVM behaviour
+
+The `RETURNDATASIZE` opcode gets size of output data from the previous call from the current environment.
+
+### Circuit behaviour
+
+1. Construct call context table in rw table
+2. Do busmapping lookup for call context last call return data length read operation
+3. Do busmapping lookup for stack write operation
```
Comment (CPerezz) [inline specs/opcode/3dRETURNDATASIZE.md:13]:
```suggestion
2. Do a busmapping lookup for CallContext last Call's ReturnDataLength  read.
3. Do a busmapping lookup for a stack write operation corresponding to the RETURNDATASIZE result.
```

File: specs/opcode/3dRETURNDATASIZE.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+1. opId == 0x3D
+2. State transition:
+   - gc + 2
+   - stack_pointer - 1
+   - pc + 1
+   - gas + 2
+3. Lookups: 2
+   - `u64` is in the rw table {call context, call ID, call data length}
+   - `u64` is on top of stack
```
Comment (CPerezz) [inline specs/opcode/3dRETURNDATASIZE.md:25]: This should not be just `u64` and should actually reflect the 
```suggestion
   - ReturnDataLength is in the rw table {call context, call ID, ReturnDataLength}.
   - ReturnDataLength is on top of stack.
```

**Review by CPerezz — COMMENTED** [review]
I agree with @xiaodino 's comment on the fieldtag used in the tests. This has to be fixed.

Aside from that, some suggestions in regards of the Markdown structure.
Once addressed I'll happily approve! :smile:

**Comment by ashWhiteHat** [comment]
@xiaodino @CPerezz 
Thank you for the review.

I modified accordingly.
I would appreciate it if you could confirm.
Thank you.

File: src/zkevm_specs/evm/execution/returndatasize.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    opcode = instruction.opcode_lookup(True)
+    instruction.constrain_equal(opcode, Opcode.RETURNDATASIZE)
+
+    # check [rw_table, call_context] table for last call return data length and compare
+    # against stack top after push.
+    instruction.constrain_equal(
+        instruction.call_context_lookup(CallContextFieldTag.LastCalleeReturnDataLength),
+        # NOTE: We can replace this with N_BYTES_WORD if we reuse the 32 byte RLC constraint in
+        # all places. See: https://github.com/privacy-scaling-explorations/zkevm-specs/issues/101
+        instruction.rlc_to_fq(instruction.stack_push(), N_BYTES_MEMORY_ADDRESS),
```
Comment (CPerezz) [inline src/zkevm_specs/evm/execution/returndatasize.py:18]: I'm wondering if it's correct to use `N_BYTES_MEMORY_ADDRESS` as basically the return data size can be bigger (at least nothing seems to prevent it). Indeed since it'll be pushed to the stack could even be a 32-byte word.

I think we should use `N_BYTES_WORD` for safety here. I know it's challenging to get such a big return data size. But we should try to cover any possible scenario.

cc: @ed255 would appreciate your opinion.

**Review by CPerezz — COMMENTED** [review]
Pending the resolution of the `N_BYTES_MEMORY_ADDRESS` question.

Once Edu replies to it, we can either change it and merge or  merge directly :)

File: src/zkevm_specs/evm/execution/returndatasize.py (hunk 5 above)
Comment (ed255) [reply src/zkevm_specs/evm/execution/returndatasize.py:18]: The value `LastCalleeReturnDataLength` is written via the state transition when returning from a call: https://github.com/privacy-scaling-explorations/zkevm-specs/blob/cdc3ef2d7c27a5aa154e6dadba22954c105c223a/src/zkevm_specs/evm/instruction.py#L263-L272
The value there will always be correct.  The size of this length affects the gas cost because it requires a memory expansion.  So by reasoning on an upper bound of gas cost, I think we can conclude that it's safe to use `N_BYTES_MEMORY_ADDRESS` bytes to encode this value as it's not possible to have a bigger value unless an impossible amount of gas cost is spent.

Nevertheless, following the reasoning of https://github.com/privacy-scaling-explorations/zkevm-specs/issues/101 probably using `N_BYTES_WORD` here is ok too?

I would say to keep it as `N_BYTES_MEMORY_ADDRESS` for now so that the upper bound on the value is explicit in the code, and also to be consistent with using the minimum number of bytes for the RLC like in other opcodes.  In the future once we address #101 we can revisit all instances of the RLC usage and update them at once.

Comment (CPerezz) [reply src/zkevm_specs/evm/execution/returndatasize.py:18]: I mostly mentioned this as basically there's no "upper bound" theoretically for the return data size.
I undestand it should never be bigger than 20 bytes. But we should support what is "theoretically" possible. (That's why we have U256 for `ChainId` even we know these high IDs will never be required.

This was my only concern. And if you check correctness. If something bigger than 20 bytes is returned, the lookup will esentially fail.

Comment (ed255) [reply src/zkevm_specs/evm/execution/returndatasize.py:18]: > I mostly mentioned this as basically there's no "upper bound" theoretically for the return data size.

I think you're right on the theoretical part, but practically there's an upper bound.  The `go-ethereum` implementation is using `u64` to encode gas values, and we're following that, as it allows for a maximum block gas value that is way over what we can imagine as reasonable values.  When we constraint the gas to be 64 bits, due to the quadratic gas cost of memory expansion we can find which memory length overflows the 64 bit gas value, and the answer is that 5 bytes is enough to overflow the gas value; so in all the circuits we're using 5 bytes only to encode a memory address.

Here's how `go-ethereum` imposes on memory length: https://github.com/ethereum/go-ethereum/blob/a69d4b273d1164637e0edb2cbad2e51325b7e897/core/vm/gas_table.go#L27

I think it's safe to limit particular fields to smaller bit sizes than specified in the yellow paper as long as they don't limit our implementation in the long term and as long as other well known (like geth) implementations follow those limits as well.

So in summary, I think to argue that we should give more bytes to the memory address is the same as arguing that we should give more than 64 bits to a gas value.

And for correctness, as long as we don't try to prove blocks that have a gas limit higher than `18446744073709551615`, there shouldn't be any issue here.

Comment (CPerezz) [reply src/zkevm_specs/evm/execution/returndatasize.py:18]: As discussed offline, completelly agree. Thanks for this reasoning!! :+1:

**Review by CPerezz — APPROVED** [review]
LGTM! 
Thanks for the work @ashWhiteHat !
